Blazor updates for 3.2 Preview 3 (#17441)
dotnet/AspNetCore.Docs: aspnetcore/blazor/get-started.md

1. To create Blazor Server apps, install [Visual Studio 2019 version 16.4 or later](https://visualstudio.microsoft.com/vs/preview/) with the **ASP.NET and web development** workload.

   To create Blazor Server and Blazor WebAssembly apps, install Visual Studio 2019 16.6 Preview 2 or later with the **ASP.NET and web development** workload.

   For information on the two Blazor hosting models, *Blazor WebAssembly* and *Blazor Server*, see <xref:blazor/hosting-models>.

1. Create a new project.

1. Select **Blazor App**. Select **Next**.

1. Provide a project name in the **Project name** field or accept the default project name. Confirm the **Location** entry is correct or provide a location for the project. Select **Create**.

1. For a Blazor WebAssembly experience (Visual Studio 16.6 Preview 2 or later), choose the **Blazor WebAssembly App** template. For a Blazor Server experience (Visual Studio 16.4 or later), choose the **Blazor Server App** template. Select **Create**.

1. Press <kbd>Ctrl</kbd>+<kbd>F5</kbd> to run the app.

# [Visual Studio Code](#tab/visual-studio-code)

1. Install the [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet-core/3.1).

> 1. Optionally install the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template. Run the following command in a command shell. The [Microsoft.AspNetCore.Components.WebAssembly.Templates](https://www.nuget.org/packages/Microsoft.AspNetCore.Components.WebAssembly.Templates/) package has a preview version while Blazor WebAssembly is in preview.

Reviewer: Would this be sufficient?

```suggestion
1. Optionally install the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template by running the following command:
```

   ```dotnetcli
   dotnet new -i Microsoft.AspNetCore.Components.WebAssembly.Templates::3.2.0-preview3.20168.3
   ```

   > [!NOTE]
   > The [.NET Core SDK version 3.1.201 or later](https://dotnet.microsoft.com/download/dotnet-core/3.1) is **required** to use the 3.2 Preview 3 Blazor WebAssembly template. Confirm the installed .NET Core SDK version by running `dotnet --version` in a command shell.

1. Install [Visual Studio Code](https://code.visualstudio.com/).

> 1. Install the latest [C# for Visual Studio Code extension](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csharp) and the [JavaScript Debugger (Nightly)](https://marketplace.visualstudio.com/items?itemName=ms-vscode.js-debug-nightly) extension with `debug.javascript.usePreview` set to `true`.

Reviewer: Very nice! :smile:

# [Visual Studio for Mac](#tab/visual-studio-mac)

Blazor Server is supported in Visual Studio for Mac. Blazor WebAssembly isn't supported at this time. To build Blazor WebAssembly apps on macOS, follow the guidance on the **.NET Core CLI** tab.

1. Install [Visual Studio for Mac](https://visualstudio.microsoft.com/vs/mac/).

1. Select **File** > **New Solution** or create a **New Project**.

1. In the sidebar, select **.NET Core** > **App**.

1. Select the **Blazor Server App** template. Select **Create**.

   For information on the Blazor Server hosting model, see <xref:blazor/hosting-models>.

1. Set the **Target Framework** to **.NET Core 3.1** and select **Next**.

1. In the **Project Name** field, name the app `WebApplication1`. Select **Create**.

1. Select **Run** > **Run Without Debugging** to run the app *without the debugger*. Run the app with **Start Debugging** to run the app *with the debugger*.

If a prompt appears to trust the development certificate, trust the certificate and continue.

# [.NET Core CLI](#tab/netcore-cli/)

1. Install the [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet-core/3.1).

> 1. Optionally install the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template. Run the following command in a command shell. The [Microsoft.AspNetCore.Components.WebAssembly.Templates](https://www.nuget.org/packages/Microsoft.AspNetCore.Components.WebAssembly.Templates/) package has a preview version while Blazor WebAssembly is in preview.

Reviewer: ```suggestion
1. Optionally install the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template by running the following command:
```

[!INCLUDE[](~/includes/blazorwasm-preview-notice.md)]

To get started with Blazor, follow the guidance for your choice of tooling:

# [Visual Studio](#tab/visual-studio)

1. Install [Visual Studio 2019 version 16.4 or later](https://visualstudio.microsoft.com/vs/preview/) with the **ASP.NET and web development** workload.

   When installing Visual Studio 2019 16.6 Preview 2 or later, the .NET Core 3.1.300 SDK is automatically installed with Visual Studio. The 3.1.300 SDK automatically provides the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template to create Blazor WebAssembly apps.
   

>    If installing a version of Visual Studio earlier than version 16.6 Preview 2, optionally install the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template. Run the following command in a command shell. The [Microsoft.AspNetCore.Components.WebAssembly.Templates](https://www.nuget.org/packages/Microsoft.AspNetCore.Components.WebAssembly.Templates/) package has a preview version while Blazor WebAssembly is in preview.

Reviewer: For folks that want to get started with Blazor WebAssembly and VS I think we should recommend people use the latest preview of 16.6 and cut any instructions for working with Blazor WebAssembly and older VS versions.

>    > .NET Core SDK version 3.1.201 or later is **required** to use the 3.2 Preview 3 Blazor WebAssembly template. Confirm the installed .NET Core SDK version by running `dotnet --version` in a command shell.

Reviewer: I think you can remove this note. If you installed the correct version of VS for Blazor WebAssembly mentioned in the previous step you'll be all set.

1. Create a new project.

1. Select **Blazor App**. Select **Next**.

1. Provide a project name in the **Project name** field or accept the default project name. Confirm the **Location** entry is correct or provide a location for the project. Select **Create**.

1. For a Blazor WebAssembly experience (Visual Studio 16.6 Preview 2 or later), choose the **Blazor WebAssembly App** template. For a Blazor Server experience (Visual Studio 16.4 or later), choose the **Blazor Server App** template. Select **Create**.

>    If the Blazor WebAssembly template isn't present, reinstall the template.

Reviewer: ```suggestion
```

Author: Ok, but I ... and others ... have had trouble with this. We get reader pings on it from time to time, and I've had to reinstall the template a few times when doing VS upgrades. I'll remove it ... hopefully those behaviors have ended.

>    For information on the two Blazor hosting models, *Blazor Server* and *Blazor WebAssembly*, see <xref:blazor/hosting-models>.

Reviewer: Since the user now needs to understand the different hosting models to pick the right VS version I think this link should maybe move to the bottom of the first step.

1. Press <kbd>Ctrl</kbd>+<kbd>F5</kbd> to run the app.

> > [!NOTE]

Reviewer: At this point I think this note can be removed. We haven't used the extension in almost a year.

> If you installed the Blazor Visual Studio extension for a prior preview release of ASP.NET Core Blazor, you can uninstall the extension. Installing the Blazor templates in a command shell is now sufficient to surface the templates in Visual Studio when using a version prior to 16.6 Preview 2.

# [Visual Studio Code](#tab/visual-studio-code)

1. Install the [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet-core/3.1).

1. Optionally install the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template:

   * Install the [.NET Core 3.1.201 or later (Preview) SDK](https://dotnet.microsoft.com/download/dotnet-core/3.1).

>    * Run the following command in a command shell. The [Microsoft.AspNetCore.Components.WebAssembly.Templates](https://www.nuget.org/packages/Microsoft.AspNetCore.Components.WebAssembly.Templates/) package has a preview version while Blazor WebAssembly is in preview.

Reviewer: These sub bullets don't seem to add much. Can we just remove them? The note below already clarifies which SDK version is required.

Author: ok ... but I'll keep the cross-link to the SDK ... I'll move it.

I'll keep the preview note because we had **multiple readers** write in that they were confused by the preview version on the package.

   ```dotnetcli
   dotnet new -i Microsoft.AspNetCore.Components.WebAssembly.Templates::3.2.0-preview3.20168.3
   ```

   > [!NOTE]
   > .NET Core SDK version 3.1.201 or later is **required** to use the 3.2 Preview 3 Blazor WebAssembly template. Confirm the installed .NET Core SDK version by running `dotnet --version` in a command shell.

1. Install [Visual Studio Code](https://code.visualstudio.com/).

> 1. Install the latest [C# for Visual Studio Code extension](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csharp).

Reviewer: I think for VS Code users we should probably go ahead and tell them here to also install the JavaScript Debugger (Nightly) extension and to turn on preview debugging, like we do in the debugging topic. Might as well get them setup right from the start.

> 1. For a Blazor WebAssembly experience, execute the following command in a command shell:

Reviewer: Since Blazor WebAssembly is still in preview, I think we should provide instructions for creating a Blazor Server project first.

   ```dotnetcli
   dotnet new blazorwasm -o WebApplication1
   ```

   For a Blazor Server experience, execute the following command in a command shell:

   ```dotnetcli
   dotnet new blazorserver -o WebApplication1
   ```

   For information on the two Blazor hosting models, *Blazor Server* and *Blazor WebAssembly*, see <xref:blazor/hosting-models>.

1. Open the *WebApplication1* folder in Visual Studio Code.

> 1. For a Blazor Server project, the IDE requests that you add assets to build and debug the project. Select **Yes**.

Reviewer: With both the C# and JavaScript Debugger extensions installed, the user will get the same prompt for Blazor WebAssembly apps now, so this step no longer needs to be Blazor Server specific.

> 1. If using a Blazor Server app, run the app using the Visual Studio Code debugger. If using a Blazor WebAssembly app, execute `dotnet run` from the app's project folder.

Reviewer: Running the app using the VS Code debugger should also work for Blazor WebAssembly apps now too. There's an extra step when launching the browser if the user actually wants to debug something, but since the point here is just to get the app running, I think we can skip that in these steps.

1. Install [Visual Studio for Mac](https://visualstudio.microsoft.com/vs/mac/).

1. Select **File** > **New Solution** or create a **New Project**.

1. In the sidebar, select **.NET Core** > **App**.

1. Select the **Blazor Server App** template. Select **Create**.

   For information on the Blazor Server hosting model, see <xref:blazor/hosting-models>.

1. Set the **Target Framework** to **.NET Core 3.1** and select **Next**.

1. In the **Project Name** field, name the app `WebApplication1`. Select **Create**.

1. Select **Run** > **Run Without Debugging** to run the app *without the debugger*. Run the app with **Start Debugging** to run the app *with the debugger*.

If a prompt appears to trust the development certificate, trust the certificate and continue.

# [.NET Core CLI](#tab/netcore-cli/)

1. Install the [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet-core/3.1).

1. Optionally install the [Blazor WebAssembly](xref:blazor/hosting-models#blazor-webassembly) template:

   * Install the [.NET Core 3.1.201 or later (Preview) SDK](https://dotnet.microsoft.com/download/dotnet-core/3.1).

>    * Run the following command in a command shell. The [Microsoft.AspNetCore.Components.WebAssembly.Templates](https://www.nuget.org/packages/Microsoft.AspNetCore.Components.WebAssembly.Templates/) package has a preview version while Blazor WebAssembly is in preview.

Reviewer: Again, I think we can drop these sub bullets.

Author: Already done.

   ```dotnetcli
   dotnet new -i Microsoft.AspNetCore.Components.WebAssembly.Templates::3.2.0-preview3.20168.3
   ```

   > [!NOTE]
   > .NET Core SDK version 3.1.201 or later is **required** to use the 3.2 Preview 3 Blazor WebAssembly template. Confirm the installed .NET Core SDK version by running `dotnet --version` in a command shell.

> 1. For a Blazor WebAssembly experience, execute the following commands in a command shell:

Reviewer: Let's flip the order here as well so Blazor Server is frist.
